core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/interceptor/ResolveIdentityInterceptor.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.interceptor;
+
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SelectedAuthScheme;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.core.interceptor.SdkInternalExecutionAttribute;
+import software.amazon.awssdk.http.auth.spi.AuthSchemeOption;
+import software.amazon.awssdk.identity.spi.Identity;
+import software.amazon.awssdk.identity.spi.IdentityProvider;
+import software.amazon.awssdk.identity.spi.ResolveIdentityRequest;
+
+@SdkInternalApi
+public class ResolveIdentityInterceptor implements ExecutionInterceptor {
+
+    @Override
+    public void beforeExecution(Context.BeforeExecution context, ExecutionAttributes executionAttributes) {
+        SelectedAuthScheme<?> selectedAuthScheme =
+            executionAttributes.getAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME);
+
+        // This seems unnecessary since this interceptor won't be configured for older clients
+        if (selectedAuthScheme == null) {
+            return;
+        }
+
+        if ("smithy.api#noAuth".equals(selectedAuthScheme.authSchemeOption().schemeId())) {
+            return;
+        }
+
+        CompletableFuture<? extends Identity> identity = resolveIdentity(selectedAuthScheme);
+        executionAttributes.putAttribute(SdkInternalExecutionAttribute.IDENTITY, identity);
+    }
+
+    private CompletableFuture<? extends Identity> resolveIdentity(SelectedAuthScheme<?> selectedAuthScheme) {
+        ResolveIdentityRequest.Builder identityRequestBuilder = ResolveIdentityRequest.builder();
+
+        AuthSchemeOption authSchemeOption = selectedAuthScheme.authSchemeOption();
+        authSchemeOption.forEachIdentityProperty(identityRequestBuilder::putProperty);
+
+        IdentityProvider<?> identityProvider = selectedAuthScheme.identityProvider();
+        return identityProvider.resolveIdentity(identityRequestBuilder.build());
+    }
```

Comment:
Here we know the result of identityProvider.responseIdentity is the same type as the SelectedAuthScheme's type. Assigning it to a separate IDENTITY attribute loses that association, so you'll need to cast it later on.

Is there any way to store this identity in the select auth scheme so the compiler can make sure we don't end up with an identity that doesn't work with the selected auth scheme?

Reply:
Yeah, at one point I considered doing identityProvider.resolveIdentity in AuthSchemeInterceptor's trySelectAuthSchceme and saving the Identity directly in SelectedAuthScheme instead of the IdentityProvider. This would "work" the same, but doesn't give us the separation of concerns of keeping Identity resolution as a separate step.

The other option would be to start the SelectedAuthScheme with null Identity and then mutate (using copy builder) the SelectedAuthScheme to add the Identity after Identity resolution. I thought of avoiding null/mutating objects, so landed on the approach in this PR. But maybe the better tradeoff is to do this - hardens the matching of the types.

Reply:
Yeah, I think I would just favor resolving the identity when we select the auth scheme. I don't see a reason we would need to separate those two things in the future, and if we did we can always do it by updating the SelectedAuthScheme.

Reply:
Changed to this approach.